In [1]:
import pandas as pd
import numpy as np

In [2]:
# Q1 Load the dataset. (1 point)
df = pd.read_csv('book_data.csv')

In [3]:
# Q2 duplicates of other records based on the book titles?
df.duplicated(subset='book_title').sum()

16

In [4]:
df = df.drop_duplicates(subset='book_title')

In [5]:
# Q2 number of unique books
len(df)

1125

In [6]:
df['author_list'] = df['book_authors'].str.split('|')

In [7]:
# Q3 exactly 3 authors
df[df['author_list'].str.len()==3]['book_title'].count()

16

In [8]:
df[df['author_list'].str.len()==3]

book_authors  \
7        Michelle McNamara|Gillian Flynn|Patton Oswalt   
35   Jordan B. Peterson|Norman Doidge|Ethan Van Sciver   
99              Cynthia Hand|Jodi Meadows|Brodi Ashton   
117  Anthony Ray Hinton|Lara Love Hardin|Bryan Stev...   
124           Bill Clinton|James Patterson|David Ellis   
194     Hans Rosling|Ola Rosling|Anna Rosling Rönnlund   
199         Mia Sheridan|Lance  Greenfield|Erin Mallon   
241           Kieron Gillen|Jamie McKelvie|Matt Wilson   
373    Carys Davies|Lorenzo Luengo|Anna Llisterri Boix   
451        Masaji Ishikawa|Risa Kobayashi|Martin Brown   
542       Beatriz Williams|Lauren Willig|Karen   White   
727           Mike Mignola|Duncan Fegredo|Dave Stewart   
734         Molly Harper|Amanda Ronconi|Jonathan Davis   
754              Stephen King|Έφη Τσιρώνη|Stephen King   
842            Mike Mignola|Richard Corben|Gary Gianni   
949         Molly Harper|Amanda Ronconi|Jonathan Davis   

                                             book_desc  \
7    A masterful true crime account of the Golden S...   
35   What does everyone in the modern world need to...   
99   You may think you know the story. After a mise...   
117  In 1985, Anthony Ray Hinton was arrested and c...   
124  President Bill Clinton and bestselling novelis...   
194  Factfulness: The stress-reducing habit of only...   
199  Listening Length: 8 hours and 58 minutesA grip...   
241  When you're at the peak of your powers, there'...   
373  When widowed mule breeder Cy Bellman reads in ...   
451  \r\nAn Amazon Charts Most Read and Most Sold b...   
542  From the New York Times bestselling authors of...   
727  The basis for the upcoming feature film Hellbo...   
734  6 hrs 24 minsAnthropologist Jillian Ramsay's c...   
754  Το Καστλ Ροκ είναι μικρή πόλη και τα νέα κυκλο...   
842  Hellboy's complete story is presented for the ...   
949  6 hrs 41 minsWelcome to Mystic Bayou, a tiny t...   

               book_edition     book_format     book_isbn book_pages  \
7                       NaN           ebook  9.780000e+12  352 pages   
35   Canadian First Edition       Hardcover  9.780000e+12  448 pages   
99                      NaN       Hardcover  9.780000e+12  464 pages   
117                     NaN       Hardcover  9.780000e+12  272 pages   
124                     NaN  Kindle Edition           NaN  513 pages   
194                     NaN       Hardcover  9.780000e+12  342 pages   
199              Unabridged   Audible Audio           NaN    9 pages   
241                     NaN       Paperback  9.780000e+12  168 pages   
373                     NaN       Hardcover  9.780000e+12  160 pages   
451                     NaN  Kindle Edition           NaN  174 pages   
542                     NaN       Hardcover  9.780000e+12  408 pages   
727                     NaN       Paperback  9.780000e+12  528 pages   
734                     NaN   Audible Audio           NaN    7 pages   
754                     NaN       Paperback           NaN  176 pages   
842                     NaN       Paperback  9.780000e+12  416 pages   
949                     NaN   Audible Audio           NaN    7 pages   

     book_rating  book_rating_count  book_review_count  \
7           4.22              60784               7148   
35          4.01              34183               3653   
99          3.83               7178               2073   
117         4.63               9018               1737   
124         3.88              45492               5830   
194         4.34              25351               2758   
199         4.24               2506                382   
241         4.07               2457                320   
373         3.70               1337                322   
451         4.27              22583               2462   
542         3.92               3966                616   
727         4.66                148                 20   
734         4.05               2038                311   
754         3.77              1

In [9]:
page_num = []
for i in df['book_pages']:
    if i is np.NaN:
        page_num.append(np.NaN)
    else:
        page_num.append(int(i.replace(" pages",'')))
        
#page_num

In [10]:
# Q4
df.insert(3,'pages',page_num)
df

book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition  pages     book_format     book_isbn book_pages  \
0             NaN  435.0  Kindle Edition           NaN  435 pages   
1             NaN  393.0       Hardcover  9.780000e+12  393 pages   
2             NaN  370.0       Hardcover  9.780000e+12  370 pages   
3             NaN  449.0  Kindle Edition           NaN  449 pages   
4             NaN  334.0       Hardcover  9.780000e+12  334 pages   
...           ...    ...             ...           ...        ...   
1136          NaN  304.0       Hardcover  9.780000e+12  304 pages   
1137          NaN  417.0       Hardcover  9.780000e+12  417 pages   
1138          NaN  320.0  Kindle Edition           NaN  320 pages   
1139          NaN  419.0       Hardcover  9.780000e+12  419 pages   
1140          NaN  304.0       Hardcover  9.780000e+12  304 pages   

      book_rating  book_rating_count  book_review_count  \
0            4.33             146505              17438   
1            4.34              61357               9369   
2            4.18              62897              11948   
3            3.97             175678              20645   
4            4.47             128337              16424   
...           ...                ...                ...   
1136         3.33                359                 88   
1137         3.69               1255                259   
1138         3.40               1300                272   
1139         3.86                907                205   
1140         3.64               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url         author_list  
0     https://images.gr-assets.com/books/1501852423l...    [Kristin Hannah]  
1     https://images.gr-assets.com/books/1508879575l...   [Madeline Miller]  
2     https://images.gr-assets.com/books/1493047153l...       [Holly Black]

In [11]:
# Q5
df['pages'].median()

326.5

In [12]:
# Q6
print(df['pages'].min())
print(df['pages'].max())

0.0
992.0


In [13]:
# Q7
print(df[df['author_list'].str.len()==1]['pages'].mean())
print(df[df['author_list'].str.len()==2]['pages'].mean())
print(df[df['author_list'].str.len()==3]['pages'].mean())

333.3510292524377
274.77777777777777
277.75


In [14]:
df[df['author_list'].str.len()>1]['pages'].mean()

280.0

The average page of books with 1 author is 333.35, while those with more than 1 author is 280. So having more than 1 author result in a longer book on average.
The mean number of pages of books with 1, 2, and 3 authors are 333.35, 274.78 and 277.75.

In [15]:
# Q8
len(df[df['book_rating']>=4])

650

In [16]:
# Q9
len(df[df['book_rating']<=4.5])

1066

In [17]:
# Q10
df['floor_rating'] = df['book_rating'].apply(np.floor)
df

book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition  pages     book_format     book_isbn book_pages  \
0             NaN  435.0  Kindle Edition           NaN  435 pages   
1             NaN  393.0       Hardcover  9.780000e+12  393 pages   
2             NaN  370.0       Hardcover  9.780000e+12  370 pages   
3             NaN  449.0  Kindle Edition           NaN  449 pages   
4             NaN  334.0       Hardcover  9.780000e+12  334 pages   
...           ...    ...             ...           ...        ...   
1136          NaN  304.0       Hardcover  9.780000e+12  304 pages   
1137          NaN  417.0       Hardcover  9.780000e+12  417 pages   
1138          NaN  320.0  Kindle Edition           NaN  320 pages   
1139          NaN  419.0       Hardcover  9.780000e+12  419 pages   
1140          NaN  304.0       Hardcover  9.780000e+12  304 pages   

      book_rating  book_rating_count  book_review_count  \
0            4.33             146505              17438   
1            4.34              61357               9369   
2            4.18              62897              11948   
3            3.97             175678              20645   
4            4.47             128337              16424   
...           ...                ...                ...   
1136         3.33                359                 88   
1137         3.69               1255                259   
1138         3.40               1300                272   
1139         3.86                907                205   
1140         3.64               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url         author_list  \
0     https://images.gr-assets.com/books/1501852423l...    [Kristin Hannah]   
1     https://images.gr-assets.com/books/1508879575l...   [Madeline Miller]   
2     https://images.gr-assets.com/books/1493047153l...       [Holly Bla

In [18]:
# Q11 For each of the discretized ratings (1,2,3,4,5), what is the average number of pages? (5 points)
print(df.groupby('floor_rating')['pages'].mean())

floor_rating
2.0    285.333333
3.0    319.102845
4.0    336.114140
5.0    104.222222
Name: pages, dtype: float64


In [19]:
# Q12 For each of the discretized ratings (1,2,3,4,5), what is the average number of reviews? (5 points)
print(df.groupby('floor_rating')['book_review_count'].mean())

floor_rating
2.0    251.333333
3.0    884.114407
4.0    859.383775
5.0      0.777778
Name: book_review_count, dtype: float64


In [20]:
# Q13
df['genre_list'] = df['genres'].str.split('|')

In [21]:
df2 = df.explode('genre_list').reset_index(drop=True)
df2 = df2.drop_duplicates(subset=['book_title','genre_list'])
df2

book_authors                                          book_desc  \
0     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
2     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
3     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
5     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
...              ...                                                ...   
6490    Nisha Sharma  Winnie Mehta was never really convinced that R...   
6491    Nisha Sharma  Winnie Mehta was never really convinced that R...   
6492    Nisha Sharma  Winnie Mehta was never really convinced that R...   
6493    Nisha Sharma  Winnie Mehta was never really convinced that R...   
6495    Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition  pages     book_format     book_isbn book_pages  \
0             NaN  435.0  Kindle Edition           NaN  435 pages   
1             NaN  435.0  Kindle Edition           NaN  435 pages   
2             NaN  435.0  Kindle Edition           NaN  435 pages   
3             NaN  435.0  Kindle Edition           NaN  435 pages   
5             NaN  435.0  Kindle Edition           NaN  435 pages   
...           ...    ...             ...           ...        ...   
6490          NaN  304.0       Hardcover  9.780000e+12  304 pages   
6491          NaN  304.0       Hardcover  9.780000e+12  304 pages   
6492          NaN  304.0       Hardcover  9.780000e+12  304 pages   
6493          NaN  304.0       Hardcover  9.780000e+12  304 pages   
6495          NaN  304.0       Hardcover  9.780000e+12  304 pages   

      book_rating  book_rating_count  book_review_count  \
0            4.33             146505              17438   
1            4.33             146505              17438   
2            4.33             146505              17438   
3            4.33             146505              17438   
5            4.33             146505              17438   
...           ...                ...                ...   
6490         3.64               1322                395   
6491         3.64               1322                395   
6492         3.64               1322                395   
6493         3.64               1322                395   
6495         3.64               1322                395   

                       book_title  \
0                 The Great Alone   
1                 The Great Alone   
2                 The Great Alone   
3                 The Great Alone   
5                 The Great Alone   
...                           ...   
6490  My So-Called Bollywood Life   
6491  My So-Called Bollywood Life   
6492  My So-Called Bollywood Life   
6493  My So-Called Bollywood Life   
6495  My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fiction|Historical|Historical Fiction|Audioboo...   
2     Fiction|Historical|Historical Fiction|Audioboo...   
3     Fiction|Historical|Historical Fiction|Audioboo...   
5     Fiction|Historical|Historical Fiction|Audioboo...   
...                                                 ...   
6490  Young Adult|Contemporary|Romance|Fiction|Young...   
6491  Young Adult|Contemporary|Romance|Fiction|Young...   
6492  Young Adult|Contemporary|Romance|Fiction|Young...   
6493  Young Adult|Contemporary|Romance|Fiction|Young...   
6495  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url       author_list  \
0     https://images.gr-assets.com/books/1501852423l...  [Kristin Hannah]   
1     https://images.gr-assets.com/books/1501852423l...  [Kristin Hannah]   
2     https://images.gr-assets.com/books/1501852423l...  [Kristin Hannah]   
3     https://images.gr-assets.com/books/1501852423l...  [Kristin Hannah]   
5     https://images.gr-assets

In [22]:
# Q14: What is the average number of pages from different genres? What is the median? (10 points)
print(df2.groupby('genre_list')['pages'].mean())

genre_list
16th Century                      560.000000
20th Century                      256.000000
Abandoned                         343.923077
Abuse                             359.000000
Academic                          335.833333
                                     ...    
Young Adult                       359.770833
Young Adult Contemporary          341.250000
Young Adult Fantasy               435.363636
Young Adult Historical Fiction    320.000000
Young Adult Romance               392.000000
Name: pages, Length: 270, dtype: float64


In [23]:
print(df2.groupby('genre_list')['pages'].median())

genre_list
16th Century                      560.0
20th Century                      256.0
Abandoned                         308.0
Abuse                             349.0
Academic                          329.5
                                  ...  
Young Adult                       352.0
Young Adult Contemporary          339.5
Young Adult Fantasy               400.0
Young Adult Historical Fiction    320.0
Young Adult Romance               392.0
Name: pages, Length: 270, dtype: float64


In [24]:
# Q15: What is the average rating of books from different genres? What is the median? (10 points)
print(df2.groupby('genre_list')['book_rating'].mean())

genre_list
16th Century                      4.130000
20th Century                      4.000000
Abandoned                         3.557692
Abuse                             4.365000
Academic                          4.193333
                                    ...   
Young Adult                       3.945178
Young Adult Contemporary          3.770000
Young Adult Fantasy               3.923846
Young Adult Historical Fiction    3.900000
Young Adult Romance               3.680000
Name: book_rating, Length: 270, dtype: float64


In [25]:
print(df2.groupby('genre_list')['book_rating'].median())

genre_list
16th Century                      4.130
20th Century                      4.000
Abandoned                         3.630
Abuse                             4.300
Academic                          4.130
                                  ...  
Young Adult                       3.940
Young Adult Contemporary          3.785
Young Adult Fantasy               3.920
Young Adult Historical Fiction    3.900
Young Adult Romance               3.680
Name: book_rating, Length: 270, dtype: float64
